### PR TITLE
Add Username Prefix Enforce for DynamicFile mode

### PR DESCRIPTION
**Pull request by nnmin-aws**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

File: cmd/aws-iam-authenticator/root.go [hunk 1]
```diff
@@ -88,23 +88,25 @@ func initConfig() {
...
+		KubeconfigPregenerated:               viper.GetBool("server.kubeconfigPregenerated"),
+		StateDir:                             viper.GetString("server.stateDir"),
+		Address:                              viper.GetString("server.address"),
+		Kubeconfig:                           viper.GetString("server.kubeconfig"),
+		Master:                               viper.GetString("server.master"),
+		BackendMode:                          viper.GetStringSlice("server.backendMode"),
+		EC2DescribeInstancesQps:              viper.GetInt("server.ec2DescribeInstancesQps"),
+		EC2DescribeInstancesBurst:            viper.GetInt("server.ec2DescribeInstancesBurst"),
+		ScrubbedAWSAccounts:                  viper.GetStringSlice("server.scrubbedAccounts"),
+		DynamicFilePath:                      viper.GetString("server.dynamicfilepath"),
```
Comment (micahhausler) [inline cmd/aws-iam-authenticator/root.go:106]: for readability, can you add either a comment or blank newline above the `DynamicFile` options so the value indenting is grouped?

File: hack/dev/access-entries-username-prefix.template [hunk 2]
```diff
@@ -0,0 +1,13 @@
+{
```
Comment (micahhausler) [inline hack/dev/access-entries-username-prefix.template:1]: Is this a test fixture? If so, can we move it into a subdirectory with the name `testfixture` or something?

File: hack/e2e-dynamicfile.sh [hunk 3]
```diff
@@ -54,6 +55,75 @@ function e2e_mountfile() {
...
+  fi
+
+  #detect if run on github and allow the test account to assume role accordingly
+  if [ $CI = true ]
+      then
+          OUT=$(aws iam list-attached-user-policies --user-name awstester)
+          echo $OUT
+          if [ -z "$OUT" ]
+              then
+                  OUT=$(aws iam list-policies|grep allow-assume-role|grep Arn| awk '{print $2}')
```
Comment (micahhausler) [inline hack/e2e-dynamicfile.sh:77]: Do you need to grep/awk here? Can you just use the CLI's `--query`?

File: pkg/server/server.go [hunk 4]
```diff
@@ -384,6 +393,9 @@ func (h *handler) doMapping(identity *token.Identity) (string, []string, error)
 			if err != nil {
 				return "", nil, fmt.Errorf("mapper %s renderTemplates error: %v", m.Name(), err)
 			}
+			if m.UsernamePrefixEnforce() && ReservedPrefixExists(username, m.UsernamePrefixReserveList()) {
+				return "", nil, fmt.Errorf("mapper %s username %s has reserved prefix: %v", m.Name(), username, m.UsernamePrefixReserveList())
```
Comment (micahhausler) [inline pkg/server/server.go:397]: Can we change the text to something more informative? Right now, a user trying to debug why their user is getting authorization errors won't understand this message.

How about something along the lines of:
```
invalid username '%s' for mapper %s: username must not begin with with the following prefixes: %v

File: pkg/server/server.go [hunk 5]
```diff
@@ -384,6 +393,9 @@ func (h *handler) doMapping(identity *token.Identity) (string, []string, error)
 			if err != nil {
 				return "", nil, fmt.Errorf("mapper %s renderTemplates error: %v", m.Name(), err)
 			}
+			if m.UsernamePrefixEnforce() && ReservedPrefixExists(username, m.UsernamePrefixReserveList()) {
```
Comment (micahhausler) [inline pkg/server/server.go:396]: Can you add a unit test for this too?

File: hack/e2e-dynamicfile.sh [hunk 6]
```diff
@@ -54,6 +55,75 @@ function e2e_mountfile() {
...
+              "${policies_template}" > "${policies_json}"
+      sleep 2
+      aws iam create-role --role-name ${USERNAME_TEST_ROLE} --assume-role-policy-document file://${policies_json} 1>/dev/null
+      sleep 10
+  fi
+
+  #detect if run on github and allow the test account to assume role accordingly
+  if [ $CI = true ]
+      then
+          OUT=$(aws iam list-attached-user-policies --user-name awstester)
```
Comment (micahhausler) [inline hack/e2e-dynamicfile.sh:73]: Are we using an IAM user for these tests? Can we update the test infra to use [GitHub's OIDC auth](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-amazon-web-services) instead?

File: hack/e2e-dynamicfile.sh [hunk 7]
```diff
@@ -54,6 +55,75 @@ function e2e_mountfile() {
...
 
+function e2e_dynamicfile_username_prefix_enforce(){
+  set +e
+  RoleOutput=$(aws iam get-role --role-name ${USERNAME_TEST_ROLE} 2>/dev/null)
+
+  if [ -z "$RoleOutput" ]; then
+      sed -e "s|{{AWS_ACCOUNT}}|${AWS_ACCOUNT}|g" \
+              "${policies_template}" > "${policies_json}"
+      sleep 2
+      aws iam create-role --role-name ${USERNAME_TEST_ROLE} --assume-role-policy-document file://${policies_json} 1>/dev/null
```
Comment (micahhausler) [inline hack/e2e-dynamicfile.sh:66]: Where do we clean this role up?

File: hack/e2e-dynamicfile.sh [hunk 8]
```diff
@@ -137,18 +207,18 @@ function e2e_dynamicfile(){
           unset AWS_ACCESS_KEY_ID AWS_SECRET_ACCESS_KEY AWS_SESSION_TOKEN
           aws iam delete-role --role-name ${AWS_TEST_ROLE}
           echo "end to end testing for dynamicfile mode succeeded"
-          exit 0
 
       else
           echo "testing failed"
-          exit 1
```
Comment (micahhausler) [inline hack/e2e-dynamicfile.sh:144]: Why not exit on error?

File: hack/e2e-dynamicfile.sh [hunk 9]
```diff
@@ -54,6 +55,75 @@ function e2e_mountfile() {
...
+          OUT=$(aws iam list-attached-user-policies --user-name awstester)
+          echo $OUT
+          if [ -z "$OUT" ]
+              then
+                  OUT=$(aws iam list-policies|grep allow-assume-role|grep Arn| awk '{print $2}')
+                  if [ -z "$OUT" ]; then
+                      sed -e "s|{{AWS_ACCOUNT}}|${AWS_ACCOUNT}|g" \
+                          "${allow_assume_role_policies_template}" > "${allow_assume_role_policies_json}"
+                      sleep 2
+                      OUT=$(aws iam create-policy --policy-name allow-assume-role --policy-document file://${allow_assume_role_policies_json})
```
Comment (micahhausler) [inline hack/e2e-dynamicfile.sh:82]: Where do we clean this policy up?

**Review by micahhausler — CHANGES_REQUESTED** [review]

File: hack/dev/access-entries-username-prefix.template (hunk 2 above)
Comment (nnmin-aws) [reply hack/dev/access-entries-username-prefix.template:1]: yep, currently hack/dev/ is the directory holding all the test fixture. Would you like to move them to a subdir under hack/dev/

File: hack/e2e-dynamicfile.sh (hunk 8 above)
Comment (nnmin-aws) [reply hack/e2e-dynamicfile.sh:144]: good catch, added back exit on error

File: pkg/server/server.go (hunk 5 above)
Comment (nnmin-aws) [reply pkg/server/server.go:396]: have added it https://github.com/kubernetes-sigs/aws-iam-authenticator/pull/568/files#diff-fe7a244e18e92d8f347af17363209172e1d79224f086b0585a1bbeef3b626c4eR172

File: pkg/mapper/file/mapper.go [hunk 10]
```diff
@@ -104,3 +104,11 @@ func (m *FileMapper) Map(identity *token.Identity) (*config.IdentityMapping, err
 func (m *FileMapper) IsAccountAllowed(accountID string) bool {
 	return m.accountMap[accountID]
 }
+
+func (m *FileMapper) UsernamePrefixEnforce() bool {
+	return false
```
Comment (micahhausler) [inline pkg/mapper/file/mapper.go:109]: Why not support this in the File mapper?

File: hack/dev/access-entries-username-prefix.template (hunk 2 above)
Comment (micahhausler) [reply hack/dev/access-entries-username-prefix.template:1]: ah so this is just setup infra. After this PR, we should move all the Role creation/configuration outside of this repo and make sure its just set up and ready to use

File: cmd/aws-iam-authenticator/root.go (hunk 1 above)
Comment (micahhausler) [reply cmd/aws-iam-authenticator/root.go:111]: Mentioned offline, can you put the docstrings for each variable above the config?

File: hack/e2e-dynamicfile.sh (hunk 6 above)
Comment (micahhausler) [reply hack/e2e-dynamicfile.sh:73]: Sync'd offline, this will be a separate PR to test-infra

File: pkg/config/types.go [hunk 11]
```diff
@@ -179,4 +179,10 @@ type Config struct {
 	DynamicFilePath string
 	//use UserId for mapping, IdentityArn is not used any more when DynamicFileUserIDStrict=true
 	DynamicFileUserIDStrict bool
+	//defines if Enforce the username not to start with the reserved prefixes. if the value is true,
+	//then username with the reserved prefixes will get access denied with the error reserved prefix in username
+	DynamicFileUsernamePrefixEnforce bool
+	//defines the reserved prefixes for kubernetes username, it is:
+	//"amazon", "aws", "eks","iam","system"
```
Comment (micahhausler) [inline pkg/config/types.go:186]: This set of strings might be similar to EKS's config, but not a default anywhere in this project. Can you remove this line?

File: pkg/config/types.go [hunk 12]
```diff
@@ -179,4 +179,16 @@ type Config struct {
 	DynamicFilePath string
 	//use UserId for mapping, IdentityArn is not used any more when DynamicFileUserIDStrict=true
 	DynamicFileUserIDStrict bool
+	//reserved prefix config for username
+	ReservedPrefixConfig map[string]ReservedPrefixConfig
+}
+
+type ReservedPrefixConfig struct {
+	//backend mode defined in line 171
```
Comment (micahhausler) [inline pkg/config/types.go:187]: Nit: line numbers change often, maybe just refer to the type and field, like `Config.BackendMode`

Same comment on space between comment and text

File: pkg/config/types.go [hunk 13]
```diff
@@ -179,4 +179,16 @@ type Config struct {
 	DynamicFilePath string
 	//use UserId for mapping, IdentityArn is not used any more when DynamicFileUserIDStrict=true
 	DynamicFileUserIDStrict bool
+	//reserved prefix config for username
```
Comment (micahhausler) [inline pkg/config/types.go:182]: Nit: Go comments typically begin with a space, between the comment and the text. Also linters like https://golangci-lint.run/ will flag this.

Also, is the string key of the map the backend name?

```suggestion
	// ReservedPrefixConfig defines reserved username prefixes for each backend
```

File: pkg/mapper/dynamicfile/mapper.go [hunk 14]
```diff
@@ -18,6 +18,10 @@ func NewDynamicFileMapper(cfg config.Config) (*DynamicFileMapper, error) {
 	if err != nil {
 		return nil, err
 	}
+	if value, exists := cfg.ReservedPrefixConfig[strings.ToLower(mapper.ModeDynamicFile)]; exists {
```
Comment (micahhausler) [inline pkg/mapper/dynamicfile/mapper.go:21]: Can you document that the mapper name must be lowercased? Or do you lowercase it before storage?

File: cmd/aws-iam-authenticator/root.go [hunk 15]
```diff
@@ -115,7 +118,16 @@ func getConfig() (config.Config, error) {
...
 	}
-
+	var reservedPrefixConfig []config.ReservedPrefixConfig
+	if err := viper.UnmarshalKey("server.reservedPrefixConfig", &reservedPrefixConfig); err != nil {
+		return cfg, fmt.Errorf("invalid reserved prefix config: %v", err)
+	}
+	if len(reservedPrefixConfig) > 0 {
+		cfg.ReservedPrefixConfig = make(map[string]config.ReservedPrefixConfig)
+		for _, c := range reservedPrefixConfig {
+			cfg.ReservedPrefixConfig[c.BackendMode] = c
```
Comment (micahhausler) [inline cmd/aws-iam-authenticator/root.go:128]: Can you lowercase this key before storage?

File: hack/dev/authenticator_with_dynamicfile_mode.yaml [hunk 16]
```diff
@@ -7,6 +7,15 @@ server:
   kubeconfig: {{AUTHENTICATOR_KUBECONFIG}}
   backendmode: [ "MountedFile", "DynamicFile" ]
   dynamicfilepath: {{AUTHENTICATOR_DYNAMICFILE_PATH}}
+  reservedPrefixConfig:
+  - backendmode: dynamicfile
+    usernamePrefixEnforce: true
```
Comment (micahhausler) [inline hack/dev/authenticator_with_dynamicfile_mode.yaml:12]: Do we need this config? Can we just infer if the list is empty or not?

File: cmd/aws-iam-authenticator/root.go (hunk 15 above)
Comment (nnmin-aws) [reply cmd/aws-iam-authenticator/root.go:128]: I have removed the lowercase. It is not needed actually. in config file, 'DynamicFile` is passed in https://github.com/kubernetes-sigs/aws-iam-authenticator/pull/568/files#diff-093cb716b58164ce7c35a6cc848450d2eed7a799a55d3a2e863c0d7cd14a1b36R11

File: pkg/mapper/dynamicfile/mapper.go (hunk 14 above)
Comment (nnmin-aws) [reply pkg/mapper/dynamicfile/mapper.go]: I have removed the lowercase. It is not needed actually. in config file, 'DynamicFile` is passed in https://github.com/kubernetes-sigs/aws-iam-authenticator/pull/568/files#diff-093cb716b58164ce7c35a6cc848450d2eed7a799a55d3a2e863c0d7cd14a1b36R11

File: README.md [hunk 17]
```diff
@@ -158,7 +158,28 @@ useful if you're migrating from/to EKS and want to keep your mappings, or are
 running EKS in addition to some other AWS cluster(s) and want to have the same
 mappings in each.
 
-### 5. Set up kubectl to use authentication tokens provided by AWS IAM Authenticator for Kubernetes
+#### `DynamicFile`
+A local file specified by cfg.dynamicfilepath can serve as backend. The file
```
Comment (micahhausler) [inline README.md:162]:
```suggestion
A local file specified by cfg.dynamicfilepath can serve as the backend. The file
```

File: README.md [hunk 18]
```diff
@@ -158,7 +158,28 @@ useful if you're migrating from/to EKS and want to keep your mappings, or are
...
+content is expected to be in exactly the same format as in EKSConfigMap. Whenever
+This file content changes, authenticator will automatically reload it.  This
```
Comment (micahhausler) [inline README.md:164]:
```suggestion
this file content changes, authenticator will automatically reload it.  This 
```

File: README.md [hunk 19]
```diff
@@ -158,7 +158,28 @@ useful if you're migrating from/to EKS and want to keep your mappings, or are
 running EKS in addition to some other AWS cluster(s) and want to have the same
 mappings in each.
 
-### 5. Set up kubectl to use authentication tokens provided by AWS IAM Authenticator for Kubernetes
+#### `DynamicFile`
+A local file specified by cfg.dynamicfilepath can serve as backend. The file
+content is expected to be in exactly the same format as in EKSConfigMap. Whenever
+This file content changes, authenticator will automatically reload it.  This 
+provides more flexibility on managing the arn mappings.
```
Comment (micahhausler) [inline README.md:165]:
```suggestion
provides more flexibility on managing the ARN mappings.
```

File: README.md [hunk 20]
```diff
@@ -158,7 +158,28 @@ useful if you're migrating from/to EKS and want to keep your mappings, or are
 running EKS in addition to some other AWS cluster(s) and want to have the same
 mappings in each.
 
-### 5. Set up kubectl to use authentication tokens provided by AWS IAM Authenticator for Kubernetes
+#### `DynamicFile`
+A local file specified by cfg.dynamicfilepath can serve as backend. The file
+content is expected to be in exactly the same format as in EKSConfigMap. Whenever
```
Comment (micahhausler) [inline README.md:163]:
```suggestion
content is expected to be in exactly the same format as the EKSConfigMap. Whenever
```

File: README.md [hunk 21]
```diff
@@ -158,7 +158,28 @@ useful if you're migrating from/to EKS and want to keep your mappings, or are
...
+This file content changes, authenticator will automatically reload it.  This 
+provides more flexibility on managing the arn mappings.
+
+check https://github.com/kubernetes-sigs/aws-iam-authenticator/blob/master/hack/dev/authenticator_with_dynamicfile_mode.yaml 
+about how to configure the DynamicFile mode.
+
+run `make e2e RUNNER=kind` to play with a kind cluster with DynamicFile mode enable.
+
+### 5. how to configure reservedPrefixConfig for k8s username
```
Comment (micahhausler) [inline README.md:172]:
```suggestion
### 5. How to configure reservedPrefixConfig for Kubernetes usernames
```

File: hack/e2e-dynamicfile.sh [hunk 22]
```diff
@@ -54,6 +55,79 @@ function e2e_mountfile() {
 
 }
 
+function e2e_dynamicfile_username_prefix_enforce(){
```
Comment (micahhausler) [inline hack/e2e-dynamicfile.sh:58]: No change to this PR, but can you create an issue/AI to follow up and separate the role creation/configuration logic to a separate function? Right now the test is mixed with environment configuration

File: hack/e2e-dynamicfile.sh [hunk 23]
```diff
@@ -134,21 +208,22 @@ function e2e_dynamicfile(){
...
-          aws iam delete-role --role-name ${AWS_TEST_ROLE}
           echo "end to end testing for dynamicfile mode succeeded"
-          exit 0
 
       else
           echo "testing failed"
           exit 1
   fi
+  unset AWS_ACCESS_KEY_ID AWS_SECRET_ACCESS_KEY AWS_SESSION_TOKEN
+  aws iam delete-role --role-name ${AWS_TEST_ROLE}
```
Comment (micahhausler) [inline hack/e2e-dynamicfile.sh:218]: Why delete this role on every test? That will cause a race issue if you have 2 tests running simultaneously

**Review by micahhausler — CHANGES_REQUESTED** [review]

File: hack/e2e-dynamicfile.sh [hunk 24]
```diff
@@ -54,6 +55,79 @@ function e2e_mountfile() {
...
+  set +e
+  OUT=$(kubectl --kubeconfig=${kubectl_kubeconfig} --context="test-authenticator" get nodes 2>/var/tmp/err.txt)
+  if grep -q "Unauthorized" "/var/tmp/err.txt"; then
+      echo "end to end testing for dynamicfile mode succeeded"
+  else
+      echo "end to end testing for dynamicfile mode failed"
+      exit 1
+  fi
+  unset AWS_ACCESS_KEY_ID AWS_SECRET_ACCESS_KEY AWS_SESSION_TOKEN
+  aws iam delete-role --role-name ${USERNAME_TEST_ROLE}
```
Comment (micahhausler) [inline hack/e2e-dynamicfile.sh:128]: Can you remove the role deletion here?

File: hack/e2e-dynamicfile.sh (hunk 22 above)
Comment (nnmin-aws) [reply hack/e2e-dynamicfile.sh:58]: yep sure! will create a separate PR for it

File: README.md [hunk 25]
```diff
@@ -158,7 +158,28 @@ useful if you're migrating from/to EKS and want to keep your mappings, or are
...
+run `make e2e RUNNER=kind` to play with a kind cluster with DynamicFile mode enable.
+
+### 5. How to configure reservedPrefixConfig for Kubernetes usernames
+
+authenticator can support reserved prefix for k8s username. if the reserved prefix is
+set, then username with the reserved prefix will not be authenticated with the error
+"username must not begin with with the following prefixes:".
+
+check https://github.com/kubernetes-sigs/aws-iam-authenticator/blob/master/hack/dev/authenticator_with_dynamicfile_mode.yaml
+about how to configure the reserved prefix.
```
Comment (nckturner) [inline README.md:179]:
```suggestion
Check https://github.com/kubernetes-sigs/aws-iam-authenticator/blob/master/hack/dev/authenticator_with_dynamicfile_mode.yaml 
about how to configure the DynamicFile mode.

Run `make e2e RUNNER=kind` to play with a kind cluster with DynamicFile mode enable.

### 5. How to configure reservedPrefixConfig for Kubernetes usernames

The aws-iam-authenticator can support reserved prefix for k8s username. If the reserved prefix is
set, then the username with the reserved prefix will not be authenticated with the error
"username must not begin with with the following prefixes:".

Check https://github.com/kubernetes-sigs/aws-iam-authenticator/blob/master/hack/dev/authenticator_with_dynamicfile_mode.yaml
about how to configure the reserved prefix. 
```

File: cmd/aws-iam-authenticator/root.go [hunk 26]
```diff
@@ -103,8 +103,11 @@ func getConfig() (config.Config, error) {
 		EC2DescribeInstancesQps:           viper.GetInt("server.ec2DescribeInstancesQps"),
 		EC2DescribeInstancesBurst:         viper.GetInt("server.ec2DescribeInstancesBurst"),
 		ScrubbedAWSAccounts:               viper.GetStringSlice("server.scrubbedAccounts"),
-		DynamicFilePath:                   viper.GetString("server.dynamicfilepath"),
-		DynamicFileUserIDStrict:           viper.GetBool("server.dynamicfileUserIDStrict"),
+		//flags for dynamicfile mode
+		//DynamicFilePath: the file path containing the roleMapping and userMapping
+		DynamicFilePath: viper.GetString("server.dynamicfilepath"),
+		//DynamicFileUserIDStrict: if true, then aws UserId from sts will be used to look up the roleMapping/userMapping; or aws IdentityArn is used
+		DynamicFileUserIDStrict: viper.GetBool("server.dynamicfileUserIDStrict"),
```
Comment (nckturner) [inline cmd/aws-iam-authenticator/root.go:110]:
```suggestion
		// flags for dynamicfile mode
		// DynamicFilePath: the file path containing the roleMapping and userMapping
		DynamicFilePath: viper.GetString("server.dynamicfilepath"),
		// DynamicFileUserIDStrict: if true, then aws the UserId from sts(also known as
		// the principal ID) will be used to look up the roleMapping/userMapping; or aws
		// IdentityArn is used
		DynamicFileUserIDStrict: viper.GetBool("server.dynamicfileUserIDStrict"),
```

**Comment by nckturner** [comment]
Looks good, although it looks like a unit test is failing.

**Pull request merged**